### DISCUSSION TITLE
How to make computation faster

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#25338] [Q&A Modules: Solid mechanics]
Hello All

I wrote some codes in .C files based on source codes in the directory  `/home/pshen/Mercury2/Projects/moose/modules/tensor_mechanics/src/materials/crystal_plasticity/` and made corresponding header files in `/home/pshen/Mercury2/Projects/moose/modules/tensor_mechanics/include/materials/crystal_plasticity/`. I actually just created some new variables, made them can be output. These variables don't influence the calculation of previous codes. But that change makes computation very slow and can't converge when the load goes after 5% of strain. Do you know how to improve it, make it faster and converge? Below is my codes and input files: https://github.com/pshen20127/Test.git

Thanks
Pshen

**Comment by GiudGiud** [comment]
Hello

Can you please profile so we have an idea what's slow?
https://mooseframework.inl.gov/application_development/profiling.html
Also do you have the console log for a simulation? To get more insights on the cost drivers

Guillaume

**Comment by sapitts** [comment]
In addition to the profiling Guillaume suggested (which I suspect will in part point to the Rank2tensor operations you've added for output purposes), you may also need to loosen the tolerances within the crystal plasticity model as well. I'd recommend going back to the original `CrystalPlasticityKalidindiUpdate` class, with the new material properties you have in your input file, and explore the impact of slightly loosening the following tolerances:
- `slip_increment_tolerance`: the tolerance for the slip increment on each slip system. If you expect your system to have a lot of plastic deformation, the default value might be too small
- `resistance_tol`: the slip system resistance tolerance, for each individual slip system. If your system experiences a lot of hardening as plastic deformation occurs, you may need to increase this tolerance value

The impact of the `rtol` value (the tolerance for the stress residual delta to declare the stress state for the timestep as converged) within `ComputeMultipleCrystalPlasticityStress` may also need to be adjusted, but I would start with the two tolerance values listed above first.

Hope this helps,
Stephanie

**Reply by pshen20127** [reply]
Sorry for just noticing your message. Are slip_increment_tolerance and resistance_tol what you said as the picture below? Do you know where these 2 variables are defined and could you please recommend 2 loose values for 2 variables?
![Weixin Screenshot_20230912210006](https://github.com/idaholab/moose/assets/81499506/72e34869-5e97-44c5-8a29-2f67b3088982)
![Weixin Screenshot_20230912210149](https://github.com/idaholab/moose/assets/81499506/ffde88a8-630d-4d74-a80c-ff5fe01acb51)

**Reply by sapitts** [reply]
Yes, the `_slip_incr_tol` variable corresponds to the `slip_increment_tolerance` input file parameter, and the variable `_resistance_tol` corresponds to the `resistance_tol` input file parameter. Both input file parameters are defined in the `CrystalPlasticityStressUpdateBase` class.

Depending on the hardening behavior of your system (please see the discussion in my first post), I would recommend changing one, but not both, of the tolerance values by up-to an order-of-magnitude to see if that allows the system to converge (e.g. change the `resistance_tol` from the default 0.01 to 0.1). If this change allows the simulation to converge, then I would then try to tighten the tolerance back to 0.05 or 0.025. I do recommend completing this testing with the original Kalidindi model: the extra output Rank2tensors in your classes will make the simulation run slowly. 

Performing the profiling that @GiudGiud recommended is also a good step to identify where your simulation is being slowed down.

**Reply by pshen20127** [reply]
I will try, thank you so much.
